sharedbuffers/mapped_struct.py

```diff
@@ -1719,6 +1737,374 @@ def __str__(self):
+
+    @cython.ccall
+    @cython.locals(i=cython.Py_ssize_t, xlen=cython.Py_ssize_t,
+        pfset='proxied_frozenset', bitrep=cython.size_t, dcode=cython.char,
+        objlen=cython.longlong, itemsize=cython.uchar, offset=cython.Py_ssize_t)
+    def _frozenset_eq(self, x):
+        if isinstance(x, proxied_frozenset):
+            pfset = cython.cast(proxied_frozenset, x)
+            return self.objlist == pfset.objlist and (
+                self.bitrep_lo == pfset.bitrep_lo and
+                self.bitrep_hi == pfset.bitrep_hi)
+
+        xlen = len(self)
+        if not isinstance(x, (set, frozenset)) or xlen != len(x):
+            return False
+        elif self.objlist is None:
+            bitrep = self.bitrep_lo
+            i = 0
+
+            while bitrep != 0:
+                if (bitrep & 1) != 0 and i not in x:
+                    return False
+                bitrep >>= 1
+                i += 1
+
+            bitrep = self.bitrep_hi
+            while bitrep != 0:
+                if (bitrep & 1) != 0 and i not in x:
+                    return False
+                bitrep >>= 1
+                i += 1
+
+            return True
+
+        dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+        for i in xrange(xlen):
+            if self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None) not in x:
+                return False
+        return True
+
+    def __eq__(self, seq):
+        if isinstance(self, proxied_frozenset):
+            return cython.cast(proxied_frozenset, self)._frozenset_eq(seq)
+        else:
+            return cython.cast(proxied_frozenset, seq)._frozenset_eq(self)
+
+    @cython.ccall
+    @cython.locals(strict_subset=cython.bint, i=cython.Py_ssize_t,
+        xlen=cython.Py_ssize_t, pfset='proxied_frozenset',
+        seqlen=cython.Py_ssize_t, bitrep=cython.size_t, dcode=cython.char,
+        objlen=cython.longlong, itemsize=cython.uchar, offset=cython.Py_ssize_t,
+        dcode2=cython.char, oblen2=cython.longlong, itemsize2=cython.uchar,
+        offset2=cython.Py_ssize_t, h1=cython.ulonglong, h2=cython.ulonglong)
+    def _subset(self, seq, strict_subset):
+        i = 0
+        xlen = len(self)
+
+        if isinstance(seq, proxied_frozenset):
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                if self.objlist is None:
+                    # Both proxies use compressed representation.
+                    return (self.bitrep_lo & pfset.bitrep_lo) == self.bitrep_lo and (
+                        self.bitrep_hi & pfset.bitrep_hi) == self.bitrep_hi and (
+                        not strict_subset or self.bitlen < pfset.bitlen)
+                else:
+                    dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+                    for i in xrange(xlen):
+                        if self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None) not in pfset:
```

Comment:
You already know pfset is a bitmap. You should extract the bitmap check into a cfunc method you can call in an optimized fashion, and avoid the generic membership test for speed.

Also, if objlist hasn't a numeric typecode, it has non-numeric objects, which means it can never be a subset of a bitmap frozenset, so you could do that quick check. In essence, do a switch on the typecode and summarily return "False" for bitmap-incompatible typecodes.